### PR TITLE
CMR 4078: Added support to return Measurements and Variables in V2 facets

### PR DESCRIPTION
**Pull request by ygliuvt**
This added collection search by variables parameter (nested) and support to return Measurements and Variables in V2 facets.

### CODE REVIEW

File: common-app-lib/src/cmr/common_app/services/search/parameters/converters/nested_field.clj [hunk 1]
```diff
@@ -1,19 +1,27 @@
...
+   [cmr.common-app.services.search.query-model :as qm]
+   [cmr.transmit.kms :as kms]))
+
+(def variable-sub-fields
+  "The subfields of variable nested field."
+  [:measurement :variable])
 
 (defn get-subfield-names
   [parent-field]
   "Returns all of the subfields for the provided nested field. All nested field queries also support
```
Comment (chris-durbin) [inline common-app-lib/src/cmr/common_app/services/search/parameters/converters/nested_field.clj]: Docstring in the wrong place. Don't do a git blame. :)

File: indexer-app/src/cmr/indexer/data/concepts/variable.clj [hunk 2]
```diff
@@ -32,16 +33,35 @@
                            :latest true}
                           :variable)]
     (when-not (:deleted variable-concept)
-      (:extra-fields variable-concept))))
+      ;; associate variable association into variable concept, so we can use it to generate
+      ;; elastic document for nested variables field
+      (assoc variable-concept :variable-association variable-association))))
+
+(defn- variable-concept->elstic-doc
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/data/concepts/variable.clj]: elstic typo

File: indexer-app/src/cmr/indexer/data/index_set.clj [hunk 3]
```diff
@@ -98,6 +98,15 @@
...
    :tag-value.lowercase m/string-field-mapping})
 
+(defnestedmapping variables-mapping
+  "Defines mappings for variables."
+  {:measurement m/string-field-mapping
+   :measurement.lowercase m/string-field-mapping
+   :variable m/string-field-mapping
+   :variable.lowercase m/string-field-mapping
+   :originator-id.lowercase m/string-field-mapping
+   :variable-value.lowercase m/string-field-mapping})
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/data/index_set.clj]: What is variable value?

File: indexer-app/src/cmr/indexer/data/concepts/variable.clj [hunk 4]
```diff
@@ -32,16 +33,35 @@
...
+  "Converts the augmented variable concept into the portion going in the collection elastic document."
```
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/data/concepts/variable.clj]: Did you mean "in the **collection** elastic document"?

File: search-app/docs/api.md [hunk 5]
```diff
@@ -1435,6 +1435,18 @@ Find collections matching measurement.
...
+
+This supports option _or_.
+
+Find collections matching 'variables' param value
+
+     curl "%CMR-ENDPOINT%/collections?variables\[0\]\[variable\]=Var1"
+
+Find collections matching multiple 'variables' param values, default is :and
+
+     curl "%CMR-ENDPOINT%/collections?variables\[0\]\[variable\]=Var1&variables\[0\]\[measurement\]=M1&variables\[1\]\[variable\]=Var2"
```
Comment (chris-durbin) [inline search-app/docs/api.md]: Do we support searches for variables[0][variable]? I see the facets are using variables_h[0][variable].

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 6]
```diff
@@ -43,7 +43,7 @@
                        :sensor :data-center-h :measurement}
      :always-case-sensitive #{:echo-collection-id}
      :disallow-pattern #{:echo-collection-id}
-     :allow-or #{:attribute :science-keywords :science-keywords-h}}))
+     :allow-or #{:attribute :science-keywords :science-keywords-h :varialbes-h}}))
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: :varialbes-h typo.

Indicates we might be missing a test of the OR option.

File: search-app/src/cmr/search/services/query_execution/facets/facets_v2_results_feature.clj [hunk 7]
```diff
@@ -44,14 +46,19 @@
...
   ["Keywords" "Platforms" "Instruments" "Organizations" "Projects" "Processing levels"
-   "Output File Formats" "Reprojections"])
+   "Measurements" "Output File Formats" "Reprojections"])
 
 (defn- facet-query
   "Returns the facet query for the given facet field"
   [facet-field size query-params]
-  (if (= :science-keywords facet-field)
+  (case  facet-field
+    :science-keywords
```
Comment (chris-durbin) [inline search-app/src/cmr/search/services/query_execution/facets/facets_v2_results_feature.clj]: I was curious so I looked it up. It looks like you could can group together the expressions if they have the same result.

```
(case facet-field
  (:science-keywords :variables)
  (let [depth (hv2/get-depth-for-hierarchical-field query-params query-params facet-field)]
    (hv2/nested-facet (facets-v2-params->elastic-fields facet-field) size depth)
  ;; else
```

File: system-int-test/test/cmr/system_int_test/search/facets/collection_facets_v2_search_test.clj [hunk 8]
```diff
@@ -615,6 +640,88 @@
...
+    (variable-util/associate-by-concept-ids token
+                                            "variable2"
+                                            [{:concept-id (:concept-id coll2)}
+                                             {:concept-id (:concept-id coll3)}])
+    (variable-util/associate-by-concept-ids token
+                                            "somevariable"
+                                            [{:concept-id (:concept-id coll4)}])
+    (index/wait-until-indexed)
+    
+    ;; We only check the topic level variables facet for convenience since the whole
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/facets/collection_facets_v2_search_test.clj]: "topic" = "top" as in only the "Measurements"?

File: system-int-test/test/cmr/system_int_test/search/variable/collection_measurement_search_test.clj [hunk 9]
```diff
@@ -115,4 +118,110 @@
...
+        {:0 {:variable "Variable3"}}
+        {}
+
+        "multiple variables"
+        [coll2]
+        {:0 {:variable "Variable1"}
+         :1 {:variable "Variable2"}}
+        {}
+
+        "multiple variables option OR true"
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/variable/collection_measurement_search_test.clj]: We do have an OR test - I wonder why that typo didn't cause these tests to fail.

File: system-int-test/test/cmr/system_int_test/search/facets/facet_responses.clj [hunk 10]
```diff
@@ -122,7 +122,46 @@
...
+     [{:title "Measurement1",
+       :type "filter",
+       :applied false,
+       :count 2,
+       :links
+       {:apply
+        "http://localhost:3003/collections.json?page_size=0&include_facets=v2&variables_h%5B0%5D%5Bmeasurement%5D=Measurement1"},
+       :has_children true,
+       :children
+       [{:title "Variable1",
```
Comment (chris-durbin) [inline system-int-test/test/cmr/system_int_test/search/facets/facet_responses.clj]: Sorry if you get this comment twice - I don't think it made it the first time.

If "Measurement1" is not applied we should not display its children (so "Variable1" should not be returned).

So we should have a test that if measurement is not applied, variables are not returned, and if measurement is applied the variables below that measurement are applied, but not variables below other not applied measurements.

File: indexer-app/src/cmr/indexer/data/index_set.clj (hunk 3 above)
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/data/index_set.clj]: That is for future. For tag association, we can provide a string type value when associate a tag to a collection which can be searched by later. I am doing the same thing here for variable association, similar to `originator-id` here even though we don't have requirement for it either. We can take it off if you think we won't have string value association during variable association.

File: search-app/docs/api.md (hunk 5 above)
Comment (ygliuvt) [inline search-app/docs/api.md]: yes. Updated.

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj (hunk 6 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: Good catch. Thanks.

File: indexer-app/src/cmr/indexer/data/concepts/variable.clj (hunk 4 above)
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/data/concepts/variable.clj]: yes.

File: indexer-app/src/cmr/indexer/data/concepts/variable.clj (hunk 2 above)
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/data/concepts/variable.clj]: Fixed.

File: search-app/src/cmr/search/services/query_execution/facets/facets_v2_results_feature.clj (hunk 7 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_execution/facets/facets_v2_results_feature.clj]: This is neat. Done. Thanks.

File: system-int-test/test/cmr/system_int_test/search/facets/collection_facets_v2_search_test.clj (hunk 8 above)
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/facets/collection_facets_v2_search_test.clj]: Fixed typo.

File: common-app-lib/src/cmr/common_app/services/search/parameters/converters/nested_field.clj (hunk 1 above)
Comment (ygliuvt) [inline common-app-lib/src/cmr/common_app/services/search/parameters/converters/nested_field.clj]: :) I don't need to look. Fixed.

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj (hunk 6 above)
Comment (ygliuvt) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: It turns out this line of code is obsolete. The options check has been moved to `valid-parameter-options`. That's why all tests (including the option or test) passed. I have removed this line.

File: system-int-test/test/cmr/system_int_test/search/variable/collection_measurement_search_test.clj (hunk 9 above)
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/variable/collection_measurement_search_test.clj]: That line with the typo is obsolete. It's not used anywhere.

File: system-int-test/test/cmr/system_int_test/search/facets/facet_responses.clj (hunk 10 above)
Comment (ygliuvt) [inline system-int-test/test/cmr/system_int_test/search/facets/facet_responses.clj]: Fixed. Thanks!

File: search-app/src/cmr/search/services/parameters/parameter_validation.clj [hunk 11]
```diff
@@ -326,6 +327,23 @@
             (mapcat keys values))))
       [(msg/science-keyword-invalid-format-msg)])))
 
+(defn variables-validation
```
Comment (ygliuvt) [inline search-app/src/cmr/search/services/parameters/parameter_validation.clj]: Done.

File: indexer-app/src/cmr/indexer/data/index_set.clj (hunk 3 above)
Comment (chris-durbin) [inline indexer-app/src/cmr/indexer/data/index_set.clj]: Let's remove variable-value in that case and keep originator id.

File: search-app/src/cmr/search/services/query_execution/facets/facets_v2_results_feature.clj [hunk 12]
```diff
@@ -146,15 +151,21 @@
   [context aggs facet-fields]
   (let [base-url (collection-search-root-url context)
         query-params (parse-params (:query-string context) "UTF-8")
-        flat-facet-fields (remove #{:science-keywords} facet-fields)
-        hierarchical-facets (when ((set facet-fields) :science-keywords)
-                              (hv2/create-hierarchical-v2-facets aggs base-url query-params))
+        flat-facet-fields (remove #{:science-keywords :variables} facet-fields)
+        science-keywords-facets (when ((set facet-fields) :science-keywords)
```
Comment (ygliuvt) [inline search-app/src/cmr/search/services/query_execution/facets/facets_v2_results_feature.clj]: Done.

File: indexer-app/src/cmr/indexer/data/index_set.clj (hunk 3 above)
Comment (ygliuvt) [inline indexer-app/src/cmr/indexer/data/index_set.clj]: Done.

**Pull request merged**